### PR TITLE
Fixing commands with arguments execution

### PR DESCRIPTION
**Pull request by sunix**
For instance when using the `clone` command from the Plugin API it is trying to clone `h`, instead of `https://github.com/theia-ide/theia.git`

### CODE REVIEW

**Comment by AlexTugarev** [comment]
@sunix, @sunix, I think this is targeting the wrong issue. 

Could you first of try to change the implementation of the problematic `CommandHandler` to `execute(...args: string[]) ...`?

File: packages/core/src/common/command.ts [hunk 1]
```diff
@@ -238,7 +238,7 @@ export class CommandRegistry implements CommandService {
     async executeCommand<T>(command: string, ...args: any[]): Promise<T | undefined> {
         const handler = this.getActiveHandler(command, ...args);
         if (handler) {
-            const result = await handler.execute(...args);
+            const result = await handler.execute(args);
```
Comment (AlexTugarev) [inline packages/core/src/common/command.ts]: also, this is then inconsistent with other places where `...args` is passed.

**Comment by benoitf** [comment]
IMHO fix is in another place as well as it's like this since one year but if test is valid

**Comment by kittaakos** [comment]
> For instance when using the `clone` command from the Plugin API it is trying to clone `h`, instead of `https://github.com/theia-ide/theia.git`

I can do Git clone without any problem without the plug-in API.

File: packages/core/src/common/command.spec.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+});
+
+class EmptyContributionProvider implements ContributionProvider<CommandContribution> {
+    getContributions(recursive?: boolean | undefined): CommandContribution[] {
+        return [];
+    }
+}
+
+class ConcatCommandHandler implements CommandHandler {
+    execute(args: string[]) {
```
Comment (AlexTugarev) [inline packages/core/src/common/command.spec.ts]: please change this to `...args: string[]` and you'll have the test green as well, but without changing the `executeCommand` implementation.

**Comment by AlexTugarev** [comment]
@sunix, @benoitf, can you point to code where the `git.clone` command is executed. I really think it's just the matter of the argument being passed. maybe something like `...url` vs. `[url]`.

**Comment by benoitf** [comment]
AFAIK code is https://github.com/theia-ide/theia/blob/d40e3ddf0daf912a7ce00034d72d149ac6a22f7d/packages/git/src/browser/git-view-contribution.ts#L267-L279

**Comment by benoitf** [comment]
and plugin's code
https://github.com/eclipse/che-theia-factory-extension/blob/a6c251826521d82dbcc7b2d754b20249abf1189f/src/theia-commands.ts#L36-L42

**Comment by AlexTugarev** [comment]
@sunix, please could you just fix the command handler here?  https://github.com/theia-ide/theia/blob/d40e3ddf0daf912a7ce00034d72d149ac6a22f7d/packages/git/src/browser/git-view-contribution.ts#L270 

it should be `execute: (...args: any[]) => { `.

**Comment by sunix** [comment]
@AlexTugarev ok I will do it, i will keep the test though

**Comment by sunix** [comment]
@AlexTugarev thanks,
I updated the few commands that are also affected to the interface change https://github.com/theia-ide/theia/commit/85d741f91d6eb36efc99592bbbc359cd9d269056#diff-64960989269ab58e084f83e98fa31ed5R20

**Comment by sunix** [comment]
I still do not understand how this was working before :/

**Comment by AlexTugarev** [comment]
> I still do not understand how this was working before :/

@sunix, the command is called with undefined args by default, so `url`, etc. here

https://github.com/theia-ide/theia/blob/cc02b69dece4fb0ca06f02ff192f832fc0a97543/packages/git/src/browser/git-view-contribution.ts#L277

... had to be provided by users interactively. 

But the root cause for that is the `any` type. Consider this simplified declaration for the `CommandHandler`:
```
interface CommandHandler { 
    execute(...args: any[]): any;
}
```

This is valid `<CommandHandler>{ execute: (args: any[]) => {} }`, but `args` will be a single argument. `<CommandHandler>{ execute: (args1: any[], args2: any[]) => {} }` is valid as well. What you would want to have is probably `<CommandHandler>{ execute: (...args: any[]) => {} }` to map the arguments, right? But as the contract is too lax, errors are possible.

For primitive type like `string`, TypeScript would remind you politely about incompatible types.

**Comment by elaihau** [comment]
@sunix @AlexTugarev @benoitf 
I believe this change breaks some features such as `Run Task`:
peek 2018-12-19 15-02

Reason is the `if (args)` check in the following code always passes, and the 3rd line, i.e., `return ...;`, never gets the chance to run.
```
execute: (...args: any[]) => {
   if (args) { ... }
   return ....;
}
```

**Comment by elaihau** [comment]
A few other things are also broken, such as using CTRL+P to perform file search, and running commands that takes argument(s). 

peek 2018-12-19 15-09

// cc @marcdumais-work who is going to create a release tomorrow

**Comment by AlexTugarev** [comment]
@elaihau, thanks reporting. I'm taking care of it.

**Comment by marcdumais-work** [comment]
> @elaihau, thanks reporting. I'm taking care of it.

Thanks @AlexTugarev . Do you think thie fix will be ready in time to be merged for this afternoon's release, in ~8h? Else we might instead revert this PR and bring it back after, once fixed.

**Comment by AlexTugarev** [comment]
@marcdumais-work, the fix is already merged.

**Comment by marcdumais-work** [comment]
thanks @AlexTugarev !

**Pull request merged**